terrawrap/models/graph.py

```diff
@@ -0,0 +1,236 @@
+        :param graph: The graph to be executed.
+        :param post_graph: The list of items to be executed after the graph has been run.
+        :param prefix: The prefix an item must match to be applied.
+        """
+        self.command = command
+        self.graph = graph
+        self.graph_dict = {}
+        self.post_graph = post_graph
+        self.prefix = prefix
+        self.not_applied = set()
+        self.failures = []
+
+    # pylint: disable=too-many-locals
+    def execute_graph(self, num_parallel: int = 4, debug: bool = False, print_only_changes: bool = False):
+        """
+        Function for executing the graph. Will execute in parallel, up to the limit given in num_parallel.
+        :param num_parallel: The number of pipeline entries to run in parallel.
+        :param debug: True if Terraform debugging should be turned on.
+        :param print_only_changes: True if only directories which contained changes should be printed.
+        """
+        sources = find_source_nodes(self.graph)
+        futures_to_paths = {}
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=num_parallel) as executor:
+            for source in sources:
+                entry = self._get_or_create_entry(source)
+                if not self._has_prefix(entry):
+                    future = executor.submit(entry.no_op)
+                    futures_to_paths[future] = entry.path
+                    continue
+
+                print("Executing %s %s ..." % (entry.path, self.command))
+                future = executor.submit(entry.execute, self.command, debug=debug)
+                futures_to_paths[future] = entry.path
+
+            for future in concurrent.futures.as_completed(futures_to_paths):
+                path = futures_to_paths[future]
+                if self._get_or_create_entry(path).state != "no-op":
+                    exit_code, stdout, changes_detected = future.result()
+
+                    if print_only_changes and not changes_detected:
+                        stdout = ["No changes detected.\n"]
+
+                    print("\nFinished executing %s %s ..." % (path, self.command))
+                    print("Output:\n\n%s\n" % "".join(stdout).strip())
+
+                    if exit_code != 0:
+                        self.failures.append(path)
+
+                successors = list(self.graph.successors(path))
+                if successors:
+                    self.recursive_executor(executor, successors, num_parallel, debug, print_only_changes)
+
+        for node in self.graph:
+            item = self.graph_dict.get(node)
+            if not item:
+                self.not_applied.add(node)
+            else:
+                if item.state == "no-op":
+                    self.not_applied.add(item.path)
+        if self.not_applied:
+            print("The following directories have not been run since they are out of scope")
+            print(self.not_applied)
+
+        if self.failures:
```

Comment:
Can we add some printout of all the directories that were skipped because their dependencies failed?

Reply:
Thats in the print statement above, if its skipped because of the entries above it will be added to not applied

Reply:
Ah, okay. I wasn't sure if that was where that would end up.